File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {

    public int solution(int n, int[][] edge) {
        int answer = 0;
        
        // 인접리스트를 만들자
        List<List<Integer>> adjList = new ArrayList<>();
        for(int i = 0; i < n+1; i++) {
            adjList.add(new ArrayList<Integer>());
        }
        for(int[] e: edge) {
            adjList.get(e[0]).add(e[1]);
            adjList.get(e[1]).add(e[0]);
        }
        
        // visited, d, pq 초기화
        boolean[] visited = new boolean[n+1]; visited[1] = true;
        int[] d = new int[n+1]; Arrays.fill(d, 50000); d[0] = 0; d[1] = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]); pq.add(new int[] {1, 0});

        // 다익스트라
        while(!pq.isEmpty()) {
            int[] current = pq.remove();    
            visited[current[0]] = true;
            
            if(current[1] > d[current[0]]) continue;
            
            // d 갱신
            for(int idx: adjList.get(current[0])) {
                if(d[idx] > current[1] + 1) {
                    d[idx] = Math.min(d[idx], current[1] + 1);
                    pq.add(new int[] {idx, d[idx]});
                }
            }
        }
        
        int max = Arrays.stream(d).max().getAsInt();
        for(int i = 1; i < n+1; i++) {
            if(max == d[i]) answer++;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
def solution(n, edge):
    cost = [0 for _ in range(n+1)] # 정점 사이의 거리
    visited = [False for _ in range(n+1)] # 방문 리스트
    queue = [1] # 첫 방문 넣어놓기
    visited[1] = True # 방문함

    # 인접리스트 만들기
    adj = {}
    for i in range(1, n + 1):
        adj[i] = set()
    for e in edge:
        adj[e[0]].add(e[1])
        adj[e[1]].add(e[0])

    while queue:
        current = queue.pop(0)
        for next_node in adj[current]:
            if not visited[next_node]:
                queue.append(next_node)
                visited[next_node] = True
                cost[next_node] = cost[current] + 1
        
    return cost.count(max(cost))